src/execution.jl

```diff
@@ -117,20 +158,48 @@ end
     ci = ci_cache_lookup(ci_cache(job), src, world, world)::Union{Nothing,CodeInstance}
     if ci !== nothing
         key = (ci, cfg)
-        if haskey(cache, key)
-            obj = cache[key]
-        end
+        obj = get(cache, key, nothing)
     end
 
     # slow path: compile and link
     if obj === nothing || compile_hook[] !== nothing
-        # TODO: consider loading the assembly from an on-disk cache here
-        asm = compiler(job)
+        asm = nothing
+        path = nothing
+        ondisk_hit = false
+        @static if VERSION >= v"1.11.0-"
+            if disk_cache() # TODO: (Should we allow backends to opt out?)
+                path = cache_file(ci, cfg)
+                if path !== nothing && isfile(path)
+                    ondisk_hit = true
+                    try
+                        @debug "Loading compiled kernel for $spec from $path"
+                        asm = deserialize(path)
+                    catch ex
+                        @warn "Failed to load compiled kernel at $path" exception=(ex, catch_backtrace())
+                    end
+                end
+            end
+        end
 
+        if asm === nothing
+            asm = compiler(job)
+        end
         if obj !== nothing
             # we got here because of a *compile* hook; don't bother linking
             return obj
         end
```

Comment:
@maleadt Is there a reason for waiting for until here until we return early from the compile_hook?

E.g. should we ignore the cache if we are coming from a `compile_hook`, and ensure the `compiler(job)` is guaranteed to be executed?

Reply:
> should we ignore the cache if we are coming from a `compile_hook`, and ensure the `compiler(job)` is guaranteed to be executed?

Probably, otherwise you're not going to see e.g. the LLVM IR generated for the job (which I guess is the expected situation, even when there's a cache available).